Cabal/Distribution/Simple/Configure.hs

```diff
@@ -827,17 +866,31 @@ dependencySatisfiable
         --
         -- (However, note that internal deps don't have to be
         -- specified!)
+        --
+        -- NB: Just like the case below, we might incorrectly
+        -- determine an external internal dep is satisfiable
+        -- when it actually isn't.
         (depName `Map.member` requiredDepsMap) || isInternalDep
 
+      | isInternalDep
+      , pkg_ver `withinRange` verRange =
+        -- If a 'PackageName' is defined by an internal component,
+        -- and the user didn't specify a version range which is
+        -- incompatible with the package version, the dep is
+        -- satisfiable (and we are going to use the internal
+        -- dependency.)  Note that this doesn't mean we are
+        -- actually going to SUCCEED when we configure the package,
+        -- if UseExternalInternalDeps is True.  NB: if
+        -- the version bound fails we want to fall through to the
+        -- next case.
+        True
+
       | otherwise =
-        -- Normal operation: just look up dependency in the combined
+        -- Normal operation: just look up dependency in the
         -- package index.
-        not . null . PackageIndex.lookupDependency pkgs $ d
+        not . null . PackageIndex.lookupDependency installedPackageSet $ d
       where
-        -- NB: Prefer the INTERNAL package set
-        pkgs = PackageIndex.merge installedPackageSet internalPackageSet
-        isInternalDep = not . null
-                      $ PackageIndex.lookupDependency internalPackageSet d
+        isInternalDep = Map.member depName internalPackageSet
```

Comment:
I'm slightly struggling to see what the change in behaviour is here.

Previously we would slap the internal package set over the installed package set and do dep-based lookups in both. Now we first do a lookup for the internal and then the installed. So what is the difference?